Guide for adding Observe Only Support (#187)
upbound/upjet: docs/adding-support-for-observe-only.md

# Adding Support for Management Policies (a.k.a Observe Only) in an Upjet Based Provider

## (Re)generating a provider with Management Policies

> 
> 

Reviewer: ```suggestion
```

> Go to the repo of the provider, e.g. upbound/provider-aws, on your local.

Reviewer: ```suggestion
Check out the provider repo, e.g., upbound/provider-aws, and go to the project directory on your local machine.
```

            kingpin.FatalIfError(mgr.Start(ctrl.SetupSignalHandler()), "Cannot start controller manager")
     }
    ```

3. Generate with the latest upjet and management policies:

    ```bash
    # Bump to the latest upjet
    go get github.com/upbound/upjet@main
    go mod tidy
    ```

   Enable management policies in the generator by adding `config.WithFeaturesPackage` option:

    ```diff
    diff --git a/config/provider.go b/config/provider.go
    index 964883670..1c06a53e2 100644
    --- a/config/provider.go
    +++ b/config/provider.go
    @@ -141,6 +141,7 @@ func GetProvider() *config.Provider {
                  config.WithReferenceInjectors([]config.ReferenceInjector{reference.NewInjector(modulePath)}),
                  config.WithSkipList(skipList),
                  config.WithBasePackages(BasePackages),
    +             config.WithFeaturesPackage("internal/features"),
                  config.WithDefaultResourceOptions(
                          GroupKindOverrides(),
                          KindOverrides(),
    ```

   Generate:

    ```bash
    make generate
    ```

## Testing: Locally Running the Provider with Management Policies Enabled

> 1. Create a fresh K8s cluster

Reviewer: ```suggestion
1. Create a fresh Kubernetes cluster
```

2. Apply CRDs
3. Run the provider with `--enable-management-policies`

   You can update the `run` target in the Makefile as below

    ```diff
    diff --git a/Makefile b/Makefile
    index d529a0d6..84411669 100644
    --- a/Makefile
    +++ b/Makefile
    @@ -111,7 +111,7 @@ submodules:
     run: go.build
            @$(INFO) Running Crossplane locally out-of-cluster . . .
            @# To see other arguments that can be provided, run the command with --help instead
    -       UPBOUND_CONTEXT="local" $(GO_OUT_DIR)/provider --debug
    +       UPBOUND_CONTEXT="local" $(GO_OUT_DIR)/provider --debug --enable-management-policies
    
     # NOTE(hasheddan): we ensure up is installed prior to running platform-specific
     # build steps in parallel to avoid encountering an installation race condition.
    ```
   
    and run with: 

    ```shell
    make run
    ```

> 4. Create some resources in the AWS console and try observing them by creating a managed resource with `managementPolicy: ObserveOnly`

Reviewer: ```suggestion
4. Create some resources in the provider's management console and try observing them by creating a managed resource with `managementPolicy: ObserveOnly`. 
```

    For example:

    ```yaml
    apiVersion: rds.aws.upbound.io/v1beta1
    kind: Instance
    metadata:
      name: an-existing-dbinstance
    spec:
      managementPolicy: ObserveOnly
      forProvider:
        region: us-west-1
    ```
   

>     You should see the managed resource is ready & synced and the status is 

Reviewer: Can we add a command to help them see the information updated in the status?

     }
    ```

3. Generate with the latest upjet and management policies:

    ```bash
    # Bump to the latest upjet
    go get github.com/upbound/upjet@main
    go mod tidy
    ```

   Enable management policies in the generator by adding `config.WithFeaturesPackage` option:

    ```diff
    diff --git a/config/provider.go b/config/provider.go
    index 964883670..1c06a53e2 100644
    --- a/config/provider.go
    +++ b/config/provider.go
    @@ -141,6 +141,7 @@ func GetProvider() *config.Provider {
                  config.WithReferenceInjectors([]config.ReferenceInjector{reference.NewInjector(modulePath)}),
                  config.WithSkipList(skipList),
                  config.WithBasePackages(BasePackages),
    +             config.WithFeaturesPackage("internal/features"),
                  config.WithDefaultResourceOptions(
                          GroupKindOverrides(),
                          KindOverrides(),
    ```

   Generate:

    ```bash
    make generate
    ```

## Testing: Locally Running the Provider with Management Policies Enabled

1. Create a fresh Kubernetes cluster.

> 2. Apply all providers CRDs with `kubectl apply -f package/crds`.

Reviewer: ```suggestion
2. Apply all providers' CRDs with `kubectl apply -f package/crds`.
```
